### ISSUE TITLE
Optim.Rprop issue?

### ISSUE THREAD

**Issue by gavolio** [dotnet/TorchSharp#1417]
Hi Everyone,

I am testing both PyTorch and Torchsharp with a very basic example (fitting of a single variable non-linear function). When I use Rprop in Pytorch I get a very good fitting, as expected. On the other hand, Rprop in Torchsharp yield very poor fitting, the error function is not minmized at all. I have to use the Adam optimizer to get a good fit with Torchsharp. Any advice? Thank you.

PS the PyTorch and Torchsharp are the same, line by line (same model, same number of neurons, same number of iterations, etc...)

**Comment by yueyinqiu** [comment]
maybe you shall share the code with us?

**Comment by gavolio** [comment]
//function to fit > Y = tanh(X^4 - X^3 + X^2 - X - 1)

var X = torch.arange(-2, 2, 0.04).view(-1, 1);
var Y = (X.pow(4) - X.pow(3) + X.pow(2) - X - 1).tanh();
var noise = torch.randn(X.size());
var Y_noise = Y + 0.01 * noise;
var epoch = 500;
var neurons = 20;
var learning_rates = new double[] { 1,0.1,0.01,0.001 };
var model = createModel(1, neurons, "Sigmoid");
var err = new List<double>();
var loss_fn = torch.nn.MSELoss();
var y_model = Y;
for (int l = 0; l < learning_rates.Length; l++)
{
    var optimizer = torch.optim.Rprop(model.parameters(), learning_rates[l]);
    model.zero_grad();
    for (int ep = 0; ep < epoch; ep++) 
    {
        var ym = model.forward(X);
        var error = loss_fn.forward(ym, Y_noise);
        error.backward();
        optimizer.step();
        optimizer.zero_grad();
        Debug.Print("[iteration] = " + (ep + 1).ToString() + " Error = " + error.item<float>().ToString());
        y_model = model.forward(X);
        error = loss_fn.forward(y_model, Y);
        err.Insert(ep, error.item<float>());
        
    }

**Comment by alinpahontu2912** [comment]
Hey @gavolio, I think the pr should be ready. You can check the results of your code before and after applying our fix. Thanks for reporting the issue!

**Comment by gavolio** [comment]
@alinpahontu2912 Sorry for my late answer. Thanks! I will check the results with the fix. Thanks again.
